File: src/main/java/com/mercateo/jsonschema/schema/SchemaPropertyContext.java
package com.mercateo.jsonschema.schema;

import com.mercateo.immutables.Tuple;
import com.mercateo.jsonschema.property.Property;
import com.mercateo.jsonschema.property.RawPropertyCollector;
import javaslang.collection.List;
import org.immutables.value.Value;

import java.lang.annotation.Annotation;

import static java.util.Objects.requireNonNull;

@Value.Immutable
@Tuple
public interface SchemaPropertyContext {

    PropertyChecker propertyChecker();

    Annotation[] unwrapAnnotations();

    List<RawPropertyCollector> propertyCollectors();

    default boolean isApplicable(Property property) {
        return propertyChecker().test(requireNonNull(property));
    }
}

File: src/main/java/com/mercateo/jsonschema/schema/SchemaPropertyContext.kt
package com.mercateo.jsonschema.schema

import com.fasterxml.jackson.databind.deser.impl.MethodProperty
import com.mercateo.jsonschema.property.MethodCollector
import com.mercateo.jsonschema.property.Property
import com.mercateo.jsonschema.property.RawPropertyCollector
import java.util.Objects.requireNonNull

data class SchemaPropertyContext(
        val propertyChecker: PropertyChecker = object : PropertyChecker {
            override fun test(t: Property<*, *>): Boolean {
                return true
            }
        },
        val unwrapAnnotations: List<Class<out Annotation>> = emptyList(),
        val propertyCollectors: List<RawPropertyCollector> = listOf(MethodCollector())
) {

    fun isApplicable(property: Property<*, *>): Boolean {
        return propertyChecker.test(requireNonNull(property))
    }
}